### PR TITLE
Additional code jam management utilities

### PR DESCRIPTION
**Pull request by mbaruh**
The purpose of this PR is for us to avoid manually searching the permissions of team channels, and dealing with user-based overwrites. This also allows us to quickly tell what team a participant belongs to.

Added:

- An info embed with the team the member is in. The team is decided by finding in which channel the member has overwrites.
- A Command to move a member from one team to another by changing the permissions of the appropriate team channels.
- A command to end the code jam and delete all the team channels and categories.
- A command to remove a participant from their team.

## How to review
This PR isn't as big as it may seem. Most of the diff is from moving the cog to another extension, and moving all the channel creation functions to another module. What I actually added are the few commands in the last couple of commits.

### CODE REVIEW

**Review by SebastiaanZ — APPROVED** [review]
I don't quite like the approach chosen to get additional confirmation for the command that removes the team channels. but it's not a blocker to me.

**Comment by Xithrius** [comment]
Setting this to low as the last code jam that needed these utilities occurred not to long ago, and the next code jam probably won't happen for a while.

**Review by Akarys42 — APPROVED** [review]
Seems like I reviewed this but never actually approved, lgtm!

File: bot/exts/events/code_jams/_channels.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+
+async def _get_category(guild: discord.Guild) -> discord.CategoryChannel:
+    """
+    Return a code jam category.
+
+    If all categories are full or none exist, create a new category.
+    """
+    for category in guild.categories:
+        if category.name == CATEGORY_NAME and len(category.channels) < MAX_CHANNELS:
```
[the remark this answers is not in the document]
Comment (mbaruh) [reply bot/exts/events/code_jams/_channels.py:21]: To be clear, I didn't modify this. I just moved the function. What this does is search for a category that isn't filled yet, and if there is none creates another one.

File: bot/exts/events/code_jams/_cog.py [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+
+from bot.bot import Bot
+from bot.constants import Emojis, Roles
+from bot.exts.events.code_jams import _channels
+from bot.utils.services import send_to_paste_service
+
+log = logging.getLogger(__name__)
+
+TEAM_LEADERS_COLOUR = 0x11806a
+DELETION_REACTION = "\U0001f4a5"
```
[the remark this answers is not in the document]
Comment (mbaruh) [reply bot/exts/events/code_jams/_cog.py:19]: Not familiar with it, I just went with what we use elsewhere in the bot.

File: bot/exts/events/code_jams/_cog.py [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+    """Manages the code-jam related parts of our server."""
+
+    def __init__(self, bot: Bot):
+        self.bot = bot
+
+    @commands.group(aliases=("cj", "jam"))
+    @commands.has_any_role(Roles.admins)
+    async def codejam(self, ctx: commands.Context) -> None:
+        """A Group of commands for managing Code Jams."""
+        if ctx.invoked_subcommand is None:
```
[the remark this answers is not in the document]
Comment (mbaruh) [reply bot/exts/events/code_jams/_cog.py:32]: I did not modify this, I just moved the function.

File: bot/exts/events/code_jams/_channels.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mbaruh) [reply bot/exts/events/code_jams/_channels.py:21]: And yes it being case sensitive seems fine to me, as it creates category with specific names.

File: bot/exts/events/code_jams/_cog.py [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+            await _channels.create_team_leader_channel(ctx.guild, team_leaders)
+            await ctx.send(f"{Emojis.check_mark} Created Code Jam with {len(teams)} teams.")
+
+    @codejam.command()
+    @commands.has_any_role(Roles.admins)
+    async def end(self, ctx: commands.Context) -> None:
+        """
+        Deletes all code jam channels.
+
+        Displays a confirmation message with the categories and channels to be deleted. Pressing the added reaction
```
[the remark this answers is not in the document]
Comment (mbaruh) [reply bot/exts/events/code_jams/_cog.py:84]: Doesn't make sense to me grammatically. Shouldn't it be "Display", if anything?

[the remark this answers is not in the document]
Comment (Bluenix2) [reply bot/exts/events/code_jams/_cog.py:84]: Hmm, it makes the most sense in a complete sentence
> Delete all code jam channels, displaying a confirmation with the categories and channels to be deleted.

I guess we could just leave this as-is? Or reword it to be more along the lines of
> A confirmation message is displayed with the categories and channels to be deleted.

File: bot/exts/events/code_jams/_cog.py [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+        await message.add_reaction(DELETION_REACTION)
+        try:
+            await self.bot.wait_for(
+                'reaction_add',
+                check=predicate_deletion_emoji_reaction,
+                timeout=10
+            )
+
+        except asyncio.TimeoutError:
+            await message.clear_reaction(DELETION_REACTION)
```
[the remark this answers is not in the document]
Comment (mbaruh) [reply bot/exts/events/code_jams/_cog.py:111]: When you accept it, it already adds a checkmark reaction a few lines below. Adding an "aborting" message sounds fine.

File: bot/exts/events/code_jams/_cog.py [hunk 6]
```diff
@@ -0,0 +1,235 @@
...
+                raise commands.BadArgument("You must include either a CSV file or a link to one.")
+
+            teams = defaultdict(list)
+            reader = csv.DictReader(csv_file.splitlines())
+
+            for row in reader:
+                member = ctx.guild.get_member(int(row["Team Member Discord ID"]))
+
+                if member is None:
+                    log.trace(f"Got an invalid member ID: {row['Team Member Discord ID']}")
```
Comment (Bluenix2) [inline bot/exts/events/code_jams/_cog.py:65]: Perhaps we want to keep track of this, or maybe just a bool. So that you know when running the command that some members were omitted?

File: bot/exts/events/code_jams/_cog.py [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+                check=predicate_deletion_emoji_reaction,
+                timeout=10
+            )
+
+        except asyncio.TimeoutError:
+            await message.clear_reaction(DELETION_REACTION)
+            await ctx.send("Command timed out.", reference=message)
+            return
+
+        else:
```
Comment (Bluenix2) [inline bot/exts/events/code_jams/_cog.py:115]: You return in the except, no need to have this in an `else`.

**Review by Bluenix2 — APPROVED** [review]
I found two nit-picks, feel free to disregard those unless you're changing other things.

I have not tested or ran these changes myself, at the earliest I can do that tomorrow if necessary

File: bot/exts/events/code_jams/_cog.py (hunk 6 above)
Comment (mbaruh) [reply bot/exts/events/code_jams/_cog.py:65]: This was raised in the PR which added this command, but out of scope for this one.

File: bot/exts/events/code_jams/_cog.py (hunk 7 above)
Comment (mbaruh) [reply bot/exts/events/code_jams/_cog.py:115]: I know, but I personally like the explicitness of the flow.
